### PR TITLE
fix: waitForAll with throw

### PR DESCRIPTION
**Pull request by dai-shi**
https://github.com/pmndrs/jotai/pull/358#discussion_r602177739
Sorry, my suggestion back then was wrong. We are **catching** promises, so we should **throw** what we caught. This is correct.

### CODE REVIEW

File: tests/utils/waitForAll.test.tsx [hunk 1]
```diff
@@ -85,12 +86,77 @@ it('waits for two async atoms', async () => {
 
   jest.runOnlyPendingTimers()
 
-  await findByText('num1: 1')
-  await findByText('num2: 2')
+  await findByText('num1: 1, num2: 2')
```
Comment (dai-shi) [inline tests/utils/waitForAll.test.tsx:89]: We should test with one `findByText` to avoid edge cases. (or, use `waitFor` and multiple `getByText`s.)

**Comment by Thisen** [comment]
That's fair and all - I already got 90% of this code on my fix/issue-391 branch, so maybe lets go forward with my branch and close this?

**Comment by Thisen** [comment]
You could also shave this PR down to the `throw` part and fixing the test - What you think @dai-shi ?

**Comment by dai-shi** [comment]
Let me shave this pr to fix the issue in #358, because of my misunderstanding back then.
I need to revert the atom cache usage properly.

**Review by Thisen — APPROVED** [review]
